File: tensorflowonspark/dfutil.py
# Copyright 2017 Yahoo Inc.
# Licensed under the terms of the Apache 2.0 license.
# Please see LICENSE file in the project root for terms.
"""A collection of utility functions for loading/saving TensorFlow TFRecords files as Spark DataFrames."""

from __future__ import absolute_import
from __future__ import division
from __future__ import nested_scopes
from __future__ import print_function

import tensorflow as tf
from pyspark.sql import Row
from pyspark.sql.types import ArrayType, BinaryType, DoubleType, LongType, StringType, StructField, StructType

loadedDF = {}       # Stores origin paths of loaded DataFrames (df => path)


def isLoadedDF(df):
  """Returns True if the input DataFrame was produced by the loadTFRecords() method.

  This is primarily used by the Spark ML Pipelines APIs.

  Args:
    :df: Spark Dataframe
  """
  return df in loadedDF


def saveAsTFRecords(df, output_dir):
  """Save a Spark DataFrame as TFRecords.

  This will convert the DataFrame rows to TFRecords prior to saving.

  Args:
    :df: Spark DataFrame
    :output_dir: Path to save TFRecords
  """
  tf_rdd = df.rdd.mapPartitions(toTFExample(df.dtypes))
  tf_rdd.saveAsNewAPIHadoopFile(output_dir, "org.tensorflow.hadoop.io.TFRecordFileOutputFormat",
                                keyClass="org.apache.hadoop.io.BytesWritable",
                                valueClass="org.apache.hadoop.io.NullWritable")


def loadTFRecords(sc, input_dir, binary_features=[]):
  """Load TFRecords from disk into a Spark DataFrame.

  This will attempt to automatically convert the tf.train.Example features into Spark DataFrame columns of equivalent types.

  Note: TensorFlow represents both strings and binary types as tf.train.BytesList, and we need to
  disambiguate these types for Spark DataFrames DTypes (StringType and BinaryType), so we require a "hint"
  from the caller in the ``binary_features`` argument.

  Args:
    :sc: SparkContext
    :input_dir: location of TFRecords on disk.
    :binary_features: a list of tf.train.Example features which are expected to be binary/bytearrays.

  Returns:
    A Spark DataFrame mirroring the tf.train.Example schema.
  """
  import tensorflow as tf

  tfr_rdd = sc.newAPIHadoopFile(input_dir, "org.tensorflow.hadoop.io.TFRecordFileInputFormat",
                                keyClass="org.apache.hadoop.io.BytesWritable",
                                valueClass="org.apache.hadoop.io.NullWritable")

  # infer Spark SQL types from tf.Example
  record = tfr_rdd.take(1)[0]
  example = tf.train.Example()
  example.ParseFromString(bytes(record[0]))
  schema = infer_schema(example, binary_features)

  # convert serialized protobuf to tf.Example to Row
  example_rdd = tfr_rdd.mapPartitions(lambda x: fromTFExample(x, binary_features))

  # create a Spark DataFrame from RDD[Row]
  df = example_rdd.toDF(schema)

  # save reference of this dataframe
  loadedDF[df] = input_dir
  return df


def toTFExample(dtypes):
  """mapPartition function to convert a Spark RDD of Row into an RDD of serialized tf.train.Example bytestring.

  Note that tf.train.Example is a fairly flat structure with limited datatypes, e.g. tf.train.FloatList,
  tf.train.Int64List, and tf.train.BytesList, so most DataFrame types will be coerced into one of these types.

  Args:
    :dtypes: the DataFrame.dtypes of the source DataFrame.

  Returns:
    A mapPartition function which converts the source DataFrame into tf.train.Example bytestrings.
  """
  def _toTFExample(iter):

    # supported type mappings between DataFrame.dtypes and tf.train.Feature types
    float_dtypes = ['float', 'double']
    int64_dtypes = ['boolean', 'tinyint', 'smallint', 'int', 'bigint', 'long']
    bytes_dtypes = ['binary', 'string']
    float_list_dtypes = ['array<float>', 'array<double>']
    int64_list_dtypes = ['array<boolean>', 'array<tinyint>', 'array<smallint>', 'array<int>', 'array<bigint>', 'array<long>']

    def _toTFFeature(name, dtype, row):
      feature = None
      if dtype in float_dtypes:
        feature = (name, tf.train.Feature(float_list=tf.train.FloatList(value=[row[name]])))
      elif dtype in int64_dtypes:
        feature = (name, tf.train.Feature(int64_list=tf.train.Int64List(value=[row[name]])))
      elif dtype in bytes_dtypes:
        feature = (name, tf.train.Feature(bytes_list=tf.train.BytesList(value=[str(row[name])])))
      elif dtype in float_list_dtypes:
        feature = (name, tf.train.Feature(float_list=tf.train.FloatList(value=row[name])))
      elif dtype in int64_list_dtypes:
        feature = (name, tf.train.Feature(int64_list=tf.train.Int64List(value=row[name])))
      else:
        raise Exception("Unsupported dtype: {0}".format(dtype))
      return feature

    results = []
    for row in iter:
      features = dict([_toTFFeature(name, dtype, row) for name, dtype in dtypes])
      example = tf.train.Example(features=tf.train.Features(feature=features))
      results.append((bytearray(example.SerializeToString()), None))
    return results

  return _toTFExample


def infer_schema(example, binary_features=[]):
  """Given a tf.train.Example, infer the Spark DataFrame schema (StructFields).

  Note: TensorFlow represents both strings and binary types as tf.train.BytesList, and we need to
  disambiguate these types for Spark DataFrames DTypes (StringType and BinaryType), so we require a "hint"
  from the caller in the ``binary_features`` argument.

  Args:
    :example: a tf.train.Example
    :binary_features: a list of tf.train.Example features which are expected to be binary/bytearrays.

  Returns:
    A DataFrame StructType schema
  """
  def _infer_sql_type(k, v):
    # special handling for binary features
    if k in binary_features:
      return BinaryType()

    if v.int64_list.value:
      result = v.int64_list.value
      sql_type = LongType()
    elif v.float_list.value:
      result = v.float_list.value
      sql_type = DoubleType()
    else:
      result = v.bytes_list.value
      sql_type = StringType()

    if len(result) > 1:             # represent multi-item tensors as Spark SQL ArrayType() of base types
      return ArrayType(sql_type)
    else:                           # represent everything else as base types (and empty tensors as StringType())
      return sql_type

  return StructType([StructField(k, _infer_sql_type(k, v), True) for k, v in sorted(example.features.feature.items())])


def fromTFExample(iter, binary_features=[]):
  """mapPartition function to convert an RDD of serialized tf.train.Example bytestring into an RDD of Row.

  Note: TensorFlow represents both strings and binary types as tf.train.BytesList, and we need to
  disambiguate these types for Spark DataFrames DTypes (StringType and BinaryType), so we require a "hint"
  from the caller in the ``binary_features`` argument.

  Args:
    :iter: the RDD partition iterator
    :binary_features: a list of tf.train.Example features which are expected to be binary/bytearrays.

  Returns:
    An array/iterator of DataFrame Row with features converted into columns.
  """
  # convert from protobuf-like dict to DataFrame-friendly dict
  def _get_value(k, v):
    # special handling for binary features
    if k in binary_features:
      return bytearray(v.bytes_list.value[0])

    if v.int64_list.value:
      result = v.int64_list.value
    elif v.float_list.value:
      result = v.float_list.value
    else:
      result = v.bytes_list.value

    if len(result) > 1:         # represent multi-item tensors as python lists
      return list(result)
    elif len(result) == 1:      # extract scalars from single-item tensors
      return result[0]
    else:                       # represent empty tensors as python None
      return None

  results = []
  for record in iter:
    example = tf.train.Example()
    example.ParseFromString(bytes(record[0]))       # record is (bytestr, None)
    d = {k: _get_value(k, v) for k, v in sorted(example.features.feature.items())}
    row = Row(**d)
    results.append(row)

  return results

File: src/main/scala/com/yahoo/tensorflowonspark/DFUtil.scala
/**
  * Copyright 2018 Yahoo Inc.
  * Licensed under the terms of the Apache 2.0 license.
  * Please see LICENSE file in the project root for terms.
  */
package com.yahoo.tensorflowonspark

import org.apache.hadoop.io.BytesWritable
import org.apache.spark.SparkContext
import org.apache.spark.sql.{DataFrame, Row, SparkSession}
import org.apache.spark.sql.types._
import org.tensorflow.example.Feature.KindCase
import org.tensorflow.example._
import com.google.protobuf.ByteString

import scala.collection.mutable.ListBuffer

/**
  * Helper object for loading TFRecords into DataFrames.
  */
object DFUtil {

  /**
    * This will attempt to automatically convert the tf.train.Example features into Spark DataFrame columns of equivalent types.
    *
    * Note: TensorFlow represents both strings and binary types as tf.train.BytesList, and we need to disambiguate these types
    * for Spark DataFrame DTypes (StringType and BinaryType), so we require a "hint" from the caller in the `binary_features`
    * argument.
    * @param sc SparkContext
    * @param inputDir path to TFRecords on disk.
    * @param schemaHint Spark DataFrame schema used as a hint when inferring the schema.  This can be either a partial schema
    *                   or a fully-specified schema.
    * @return a Spark DataFrame representing the TFRecords from the inputDir.
    */
  def loadTFRecords(inputDir: String, schemaHint: StructType=new StructType())(implicit sc: SparkContext): DataFrame = {
    // load as RDD[Example]
    val rawRDD = sc.newAPIHadoopFile(inputDir,
                                    classOf[org.tensorflow.hadoop.io.TFRecordFileInputFormat],
                                    classOf[org.apache.hadoop.io.BytesWritable],
                                    classOf[org.apache.hadoop.io.NullWritable])

    // infer Spark SQL types from tf.Example
    val bytesRDD = rawRDD.map { case (bytes,_) => bytes.getBytes }
    val example = Example.parseFrom(bytesRDD.take(1).head)
    val schema = inferSchema(example, schemaHint)

    val exampleRDD = rawRDD.map { case (bytes, _) => Example.parseFrom(bytes.getBytes) }

    // convert RDD[Example] to RDD[Row]
    val rdd = exampleRDD.map(fromTFExample(_, schema))
    val spark = SparkSession.builder.getOrCreate()

    // create a DataFrame using the inferred schema
    spark.createDataFrame(rdd, schema)
  }

  /**
    * Given a TensorFlow Example, infers the equivalent Spark DataFrame schema, allowing for "hints" from the caller
    * to disambiguate types, e.g. BYTES_LIST can map to either StringType and BinaryType, INT64_LIST can map to either
    * IntegerType or LongType, etc.
    *
    * @param example TensorFlow Example.
    * @param schemaHint Spark schema StructType.  Any fields in this hint, will be returned with the hinted type.  Any fields
    *                   not present in this hint will be inferred with "best guesses".
    * @return a fully-specified Spark schema StructType.
    */
  def inferSchema(example: Example, schemaHint: StructType): StructType = {
    import scala.collection.JavaConversions._

    // convert schemaHint to a map for easier lookup
    val ftypes = schemaHint.map(f => (f.name, f.dataType)).toMap

    def _inferSchemaType(name: String, f: Feature): DataType = {
      // if user-hint provided, just return it
      if (ftypes.contains(name)) return ftypes(name)

      // else infer as best we can
      val dtype = f.getKindCase match {
        case KindCase.BYTES_LIST => StringType    // prefer StringType, since it's most likely?
        case KindCase.INT64_LIST => LongType
        case KindCase.FLOAT_LIST => FloatType
        case KindCase.KIND_NOT_SET => throw new Exception(s"Unsupported TFRecord type: ${f.getKindCase}")
      }

      // check if there are multiple values, i.e. array
      val isArray = dtype match {
        case BinaryType => f.getBytesList.getValueList.size > 1
        case StringType => f.getBytesList.getValueList.size > 1
        case LongType => f.getInt64List.getValueList.size > 1
        case FloatType => f.getFloatList.getValueList.size > 1
      }

      if (isArray) ArrayType(dtype) else dtype
    }

    val fmap = example.getFeatures.getFeatureMap.toMap

    // First, try to preserve order of hinted fields (e.g. if a full schema was supplied)
    val hintedFields = schemaHint.map(field =>
      StructField(field.name, _inferSchemaType(field.name, fmap(field.name)), field.nullable))
    val hintedFieldNames = schemaHint.map(field => field.name)

    // Next, infer remaining fields.  Note that fmap does not declare/preserve any order
    val remaining = fmap.filterNot { case (name, feature) => hintedFieldNames.contains(name) }
    val remainingFields = remaining.toSeq.map { case (name, feature) =>
        StructField(name, _inferSchemaType(name, feature), nullable=true)
    }

    StructType(hintedFields ++ remainingFields)
  }

  /**
    * Converts a TensorFlow example to an equivalent Spark DataFrame Row.
    *
    * @param example TensorFlow Example
    * @param schema Fully-specified Spark schema
    * @return Spark DataFrame Row
    */
  def fromTFExample(example: Example, schema: StructType): Row = {
    import scala.collection.JavaConversions._
    val fields = schema.fields
    val fmap = example.getFeatures.getFeatureMap.toMap

    def _getValue(field: StructField): Any = {
      val f = fmap(field.name)
      f.getKindCase match {
        case KindCase.BYTES_LIST =>
          val byteList = f.getBytesList
          byteList.getValueCount match {
            case 0 => require(field.nullable, s"Field ${field.name} is not nullable"); null
            case 1 => field.dataType match {
              case BinaryType => byteList.getValue(0).toByteArray
              case BooleanType => byteList.getValue(0).byteAt(0) > 0
              case StringType => byteList.getValue(0).toStringUtf8
              case _ => throw new Exception(s"Unsupported BYTES_LIST type conversion to ${field.dataType}")
            }
            case _ => field.dataType match {
              case ArrayType(BinaryType,_) => byteList.getValueList.map(_.toByteArray).toList
              case ArrayType(BooleanType,_) => byteList.getValueList.map(_.byteAt(0) > 0).toList
              case ArrayType(StringType,_) => byteList.getValueList.map(_.toStringUtf8).toList
            }
          }
        case KindCase.INT64_LIST =>
          val int64List = f.getInt64List
          int64List.getValueCount match {
            case 0 => require(field.nullable, s"Field ${field.name} is not nullable"); null
            case 1 => field.dataType match {
              case BooleanType => int64List.getValue(0) != 0
              case IntegerType => int64List.getValue(0).toInt
              case LongType => int64List.getValue(0)
              case FloatType => int64List.getValue(0).toFloat
              case _ => throw new Exception(s"Unsupported INT64_LIST type conversion to ${field.dataType}")
            }
            case _ => field.dataType match {
              case ArrayType(BooleanType,_) => int64List.getValueList.map(_ != 0).toList
              case ArrayType(IntegerType,_) => int64List.getValueList.map(_.toInt).toList
              case ArrayType(LongType,_) => int64List.getValueList.toList
              case ArrayType(FloatType,_) => int64List.getValueList.map(_.toFloat).toList
              case _ => throw new Exception(s"Unsupported INT64_LIST conversion to ${field.dataType}")
            }
          }
        case KindCase.FLOAT_LIST =>
          val floatList = f.getFloatList
          floatList.getValueCount match {
            case 0 => require(field.nullable, s"Field ${field.name} is not nullable"); null
            case 1 => field.dataType match {
              case FloatType => floatList.getValue(0)
              case DoubleType => floatList.getValue(0).toDouble
              case _ => throw new Exception(s"Unsupported FLOAT_LIST conversion to ${field.dataType}")
            }
            case _ => field.dataType match {
              case ArrayType(FloatType, _) => floatList.getValueList.toList
              case ArrayType(DoubleType, _) => floatList.getValueList.map(_.toDouble).toList
              case _ => throw new Exception(s"Unsupported FLOAT_LIST conversion to ${field.dataType}")
            }
          }
        case KindCase.KIND_NOT_SET =>
          throw new Exception(s"Unsupported TFRecord type: ${f.getKindCase}")
      }
    }

    val cols = fields.map(_getValue)
    Row.fromSeq(cols)
  }

  def saveAsTFRecords(df: DataFrame, outputDir: String): Unit = {
    val rdd = df.rdd.mapPartitions(toTFExample(df.schema))
    val pairRDD = rdd.map(e => (new BytesWritable(e.toByteArray), null))
    pairRDD.saveAsNewAPIHadoopFile(outputDir,
                                  classOf[org.apache.hadoop.io.BytesWritable],
                                  classOf[org.apache.hadoop.io.NullWritable],
                                  classOf[org.tensorflow.hadoop.io.TFRecordFileOutputFormat])
  }

  def toTFExample(schema: StructType)(iter: Iterator[Row]): Iterator[Example] = {
    import scala.collection.JavaConversions._

    val dtypes = schema.fields.zipWithIndex.map { case (f,i) => (f.name, f.dataType, i) }

    def _toTFFeature(dtype: DataType, index: Int, row: Row): Feature = {
      val feature = dtype match {
        case BinaryType =>
          val bytes = row.getAs[Array[Byte]](index)
          Feature.newBuilder.setBytesList(BytesList.newBuilder.addValue(ByteString.copyFrom(bytes)))
        case BooleanType =>
          Feature.newBuilder.setInt64List(Int64List.newBuilder.addValue(if (row.getBoolean(index)) 1L else 0L))
        case FloatType =>
          Feature.newBuilder.setFloatList(FloatList.newBuilder.addValue(row.getFloat(index)))
        case DoubleType =>
          Feature.newBuilder.setFloatList(FloatList.newBuilder.addValue(row.getDouble(index).toFloat))
        case LongType =>
          Feature.newBuilder.setInt64List(Int64List.newBuilder.addValue(row.getLong(index)))
        case IntegerType =>
          Feature.newBuilder.setInt64List(Int64List.newBuilder.addValue(row.getInt(index)))
        case StringType =>
          Feature.newBuilder.setBytesList(BytesList.newBuilder.addValue(ByteString.copyFromUtf8(row.getString(index))))
        case ArrayType(BinaryType, _) =>
          val builder = BytesList.newBuilder
          row.getList[Array[Byte]](index).foreach(b => builder.addValue(ByteString.copyFrom(b)))
          Feature.newBuilder.setBytesList(builder)
        case ArrayType(BooleanType, _) =>
          val builder = Int64List.newBuilder
          row.getList[java.lang.Boolean](index).foreach(b => builder.addValue(if (b) 1L else 0L))
          Feature.newBuilder.setInt64List(builder)
        case ArrayType(IntegerType, _) =>
          val builder = Int64List.newBuilder
          row.getList[java.lang.Integer](index).foreach(i => builder.addValue(i.longValue))
          Feature.newBuilder.setInt64List(builder)
        case ArrayType(LongType, _) =>
          val arr = row.getList[java.lang.Long](index)
          Feature.newBuilder.setInt64List(Int64List.newBuilder.addAllValue(arr))
        case ArrayType(FloatType, _) =>
          val arr = row.getList[java.lang.Float](index)
          Feature.newBuilder.setFloatList(FloatList.newBuilder.addAllValue(arr))
        case ArrayType(DoubleType, _) =>
          val builder = FloatList.newBuilder
          row.getList[java.lang.Double](index).foreach(d => builder.addValue(d.floatValue))
          Feature.newBuilder.setFloatList(builder)
        case ArrayType(StringType, _) =>
          val arr = row.getList[String](index).toList.map(s => ByteString.copyFrom(s.getBytes))
          Feature.newBuilder.setBytesList(BytesList.newBuilder.addAllValue(arr))
      }
      feature.build()
    }

    var result = ListBuffer.empty[Example]
    for (row <- iter) {
      val fbuilder = Features.newBuilder()
      dtypes.foreach { case (name, dtype, index) =>
        fbuilder.putFeature(name, _toTFFeature(dtype, index, row))
      }
      val example = Example.newBuilder.setFeatures(fbuilder.build()).build()

      result += example
    }

    result.iterator
  }
}
